### PR TITLE
CLIENT-6284 | Adding reconnectin and reconnected events and states

### PR DESCRIPTION
**Pull request by charliesantos**
**Contributing to Twilio**

> All third party contributors acknowledge that any contributions they provide will be made under the same open source license that the open source project is provided under.

- [x] I acknowledge that all my contributions will be made under the project's license.

### JIRA link(s):

- [CLIENT-6284](https://issues.corp.twilio.com/browse/CLIENT-6284)

### Description

- Add new `reconnecting` event on transition to new `reconnecting` state
- Add new `reconnected` event on transition from `reconnecting` to `open`
- Publishing events to insights on reconnection and on reconnected
- Adding ice_restart_enabled on insights default payload

### Before review
* [x] Updated CHANGELOG.md if necessary
* [x] Added unit tests if necessary
* [ ] Updated affected documentation
* [ ] Verified locally with `npm run build:release`
* [x] Manually sanity tested running locally
* [x] Ready for review

### Before merge
* [ ] Got one or more +1s
* [ ] Squashed erroneous commits if necessary
* [ ] Re-tested if necessary

### CODE REVIEW

File: lib/twilio/connection.ts [hunk 1]
```diff
@@ -268,18 +268,23 @@ class Connection extends EventEmitter {
...
           return;
         }
 
+        const mediaFailedError = { code: 53405, message: 'Media connection failed.' };
+
         this._log.warn('ICE Connection disconnected.');
-        // Stop existing loops if this warning is emitted multiple times
-        this._stopIceRestarts();
+        this._publisher.info('connection', 'reconnecting', null, this);
+        this._publisher.warn('connection', 'error', mediaFailedError, this);
```
Comment (ryan-rowland) [inline lib/twilio/connection.ts]: Should these events be ordered the other way around?

File: lib/twilio/device.ts [hunk 2]
```diff
@@ -784,6 +784,7 @@ class Device extends EventEmitter {
   private _createDefaultPayload = (connection?: Connection): Record<string, any> => {
     const payload: Record<string, any> = {
       dscp: !!this.options.dscp,
+      ice_restart_enabled: this.options.enableIceRestart,
```
Comment (ryan-rowland) [inline lib/twilio/device.ts]: This needs to be inside the payload, this is the headers section.

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,3 +1,20 @@
...
+This feature, when `enableIceRestart` is enabled, allows for detecting when media connection fails which will trigger automatic media reconnection, and for detecting when media connection is restored.
+
+#### New Events
+* `Connection.on('reconnecting', handler(connection, error))` - raised when media connection fails and automatic reconnection has been started. Media connection failure triggers when there is no media flowing for approximately more than 3 seconds. During this period, `Connection.status()` will be set to `reconnecting`;
+  * Parameters passed to the handler
+    * `connection` - the current connection object
+    * `error` - Error object `{ code: 53405, message: 'Media connection failed.' }`
+* `Connection.on('reconnected', handler(connection))` - raised when media connection has been restored which is detected when media starts flowing. Once reconnected, `Connection.status()` will be set to `open`;
+  * Parameters passed to the handler
+    * `connection` - the current connection object
```
Comment (ryan-rowland) [inline CHANGELOG.md]: I don't think passing `this` in either of these events is useful, if the application is attaching something to `connection.on(event)` it already has access to `connection`.

File: lib/twilio/connection.ts (hunk 1 above)
Comment (charliesantos) [reply lib/twilio/connection.ts]: Yes, that makes sense.

File: CHANGELOG.md (hunk 3 above)
Comment (charliesantos) [reply CHANGELOG.md]: Ok, agreed.

File: lib/twilio/device.ts (hunk 2 above)
Comment (charliesantos) [reply lib/twilio/device.ts]: Sorry but I'm confused. This says createDefaultPayload and I'm seeing the flag is getting created inside the payload, below is an example. I might be missing something. Where in the code are we putting default payload?
```
{
    "publisher": "twilio-js-sdk",
    "group": "get-user-media",
    "name": "succeeded",
    "timestamp": "2019-07-22T22:22:22.087Z",
    "level": "INFO",
    "payload_type": "application/json",
    "private": false,
    "payload": {
        "dscp": true,
        "ice_restart_enabled": true,
        "platform": "WebRTC",
        "sdk_version": "1.8.0-dev",
        "selected_region": "gll",
        "temp_call_sid": "TJS9ef58afe-9eff-4cd2-a026-a24d84efcd42",
        "direction": "OUTGOING",
        "gateway": "ec2-34-203-251-109.compute-1.amazonaws.com",
        "region": "US_EAST_VIRGINIA",
        "data": {
            "audioConstraints": true
        }
    }
}
```

**Comment by charliesantos** [comment]
Will submit this against 1.8 branch

**Pull request closed without merging**